python/ray/dag/collective_node.py

```diff
@@ -0,0 +1,197 @@
+from weakref import ReferenceType
+from typing import Any, Dict, List, Union, Tuple, Optional, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    import torch
+
+import ray
+from ray.dag import (
+    DAGNode,
+    ClassNode,
+)
+from ray.dag.constants import (
+    PARENT_CLASS_NODE_KEY,
+    BIND_INDEX_KEY,
+    COLLECTIVE_GROUP_KEY,
+)
+from ray.dag.format_utils import get_dag_node_str
+from ray.util.annotations import DeveloperAPI
+from ray.util.collective.nccl_types import CollectiveOp, ReduceOp
+from ray.experimental.channel import ChannelContext
+from ray.experimental.channel.torch_tensor_nccl_channel import _init_nccl_group
+from ray.experimental.channel.torch_tensor_type import GPUCommunicator, TorchTensorType
+
+
+class CollectiveGroup:
+    """Represent metadata for a NCCL collective method."""
+
+    def __init__(
+        self,
+        input_nodes: List[DAGNode],
+        op: CollectiveOp,
+        transport: Union[str, GPUCommunicator] = TorchTensorType.NCCL,
+    ):
+        self._input_nodes: List[DAGNode] = input_nodes
+        if len(self._input_nodes) == 0:
+            raise ValueError("Expected input nodes for a collective group")
+        if len(set(self._input_nodes)) != len(self._input_nodes):
+            raise ValueError("Expected unique input nodes for a collective group")
+
+        self._actor_handles: List["ray.actor.ActorHandle"] = []
+        for input_node in self._input_nodes:
+            actor_handle = input_node._get_actor_handle()
+            assert actor_handle is not None, "Expected a actor handle"
+            self._actor_handles.append(actor_handle)
+        if len(set(self._actor_handles)) != len(self._actor_handles):
+            raise ValueError("Expected unique actor handles for a collective group")
+
+        self._op = op
+        assert isinstance(
+            self._op, ReduceOp
+        ), "Other collective ops are not implemented"
+        self._type_hint = TorchTensorType(transport=transport, _direct_return=True)
+        if isinstance(transport, GPUCommunicator):
+            assert set(transport.get_actor_handles()) == set(
+                self._actor_handles
+            ), "Expected actor handles to match the custom NCCL group"
+
+    def __str__(self) -> str:
+        return (
+            f"CollectiveGroup("
+            f"_input_nodes={self._input_nodes}, "
+            f"_actor_handles={self._actor_handles}, "
+            f"_op={self._op}, "
+            f"_type_hint={self._type_hint})"
+        )
+
+    @property
+    def actor_handles(self) -> List["ray.actor.ActorHandle"]:
+        return self._actor_handles
```

Comment:
nit: maybe just `actors` is more concise, also `self._actors`

Reply:
DAG nodes use `_get_actor_handle()`, so use `actor_handle` for consistency.